src/components/code-editor/index.jsx

```diff
@@ -1,99 +1,106 @@
+	{ tag: tags.invalid, class: 'cm-invalid' }
+]);
 
-	showError(error) {
-		clearTimeout(this.showErrorTimer);
-		if (this.errors) {
-			this.editor.operation(() => {
-				this.errors.forEach(e => e.clear());
-			});
-			this.errors.length = 0;
-		}
+/**
+ * @param {object} props
+ * @param {string} props.editorCode
+ * @param {(value: string) => void} props.onInput
+ * @param {string} props.slug
+ * @param {string} [props.class]
+ */
+export default function CodeEditor(props) {
+	const editorParent = useRef(null);
+	/** @type {{ current: EditorView | null }} */
+	const editor = useRef(null);
 
-		if (!error || !error.loc) return;
+	const routeHasChanged = useRef(false);
 
-		this.showErrorTimer = setTimeout(() => {
-			this.editor.operation(() => {
-				let { left } = this.editor.cursorCoords(
-					{ line: error.loc.line - 1, ch: error.loc.column - 1 },
-					'local'
-				);
-				let ref;
-				const errorLine = (
-					<div ref={r => (ref = r)} class={style.lintError}>
-						<pre style={`padding-left:${left}px;`}>▲</pre>
-						<div>🔥 {error.message.split('\n')[0]}</div>
-					</div>
-				);
-				render(errorLine, this.scratch);
-				this.errors = [this.editor.addLineWidget(error.loc.line - 1, ref)];
-			});
-		}, 1000);
-	}
+	useEffect(() => {
+		if (props.slug || !editor.current) routeHasChanged.current = true;
+	}, [props.slug]);
 
-	componentDidMount() {
-		let { spaces, value, tabSize } = this.props;
+	useEffect(() => {
+		if (routeHasChanged.current === false) return;
+		routeHasChanged.current = false;
 
-		this.editor = codemirror(this.base, {
-			value: String(value || ''),
-			mode: 'jsx',
-			theme: 'one-dark',
-			lineNumbers: true,
-			indentWithTabs: !spaces,
-			tabSize: tabSize || 2,
-			indentUnit: spaces ? Math.round(spaces) || 2 : false,
-			showCursorWhenSelecting: true,
-			extraKeys: {
-				'Cmd-/': 'toggleComment'
-			}
-		});
+		if (editor.current) {
+			editor.current.dispatch({
+				changes: { from: 0, to: editor.current.state.doc.length, insert: props.editorCode }
+			});
+			return;
+		}
 
-		this.editor.on('change', () => {
-			if (this.events === false) return;
+		const theme = EditorView.theme({}, { dark: true });
 
-			this.value = this.editor.getValue();
-			let { onInput } = this.props;
-			if (onInput) onInput(this.value);
+		const state = EditorState.create({
+			doc: props.editorCode,
+			extensions: [
+				lineNumbers(),
+				highlightActiveLine(),
+				highlightActiveLineGutter(),
+				history(),
+				indentUnit.of('\t'),
+				closeBrackets(),
+				bracketMatching(),
+				autocompletion(),
+				javascript({ jsx: true }),
+				keymap.of([indentWithTab, ...defaultKeymap, ...historyKeymap]),
+				[theme, syntaxHighlighting(highlightStyle, { fallback: true })],
+				EditorView.updateListener.of(update => {
+					// Ignores changes from swapping out the editor code programmatically
+					if (isViewUpdateFromUserInput(update)) {
+						props.onInput(update.state.doc.toString());
+					}
```

Comment:
TODO: Upon further testing, this seems to be a race condition. Any input triggers two events, one in which `isViewUpdateFromUserInput()` returns `true`, the second, `false`. Which has the updated doc string is inconsistent.